src/package-hoister.js

```diff
@@ -242,7 +242,17 @@ export default class PackageHoister {
           continue;
         }
 
-        const isMarkedAsOptional = depinfo.pkg._reference && depinfo.pkg._reference.optional && this.ignoreOptional;
+        const depRef = depinfo.pkg._reference;
+
+        let isMarkedAsOptional = depRef && depRef.optional && this.ignoreOptional;
```

Comment:
We should be able to reduce this to:
```
let isMarkedAsOptional = depRef && depRef.optional && this.ignoreOptional && !(info.isRequired && depRef.hint !== 'optional');
```

What do you think? The comment you added should stay though, it is great!